Add Replicated Table GUC Fallback

thedanhoffman opened the pull request:
This adds a GUC which defaults any DML operation on a replicated table to fall
back to Planner. Behavior has been unstable and a more conservative default has
been chosen. This also updates the corresponding tests to change the default
behavior to allow Orca to generate plans with DML operations on replicated
tables (the net change is adding a SET to these files and a RESET at the end).

vraghavan78 commented on src/backend/gpopt/config/CConfigParamMapping.cpp:
```diff
@@ -290,6 +290,8 @@ CConfigParamMapping::SConfigMappingElem CConfigParamMapping::m_elements[] = {
 	 GPOS_WSZ_LIT(
 		 "Explore a nested loop join even if a hash join is possible")},
+	{EopttraceEnableReplicatedTable, &optimizer_enable_replicated_table, false,
+	 GPOS_WSZ_LIT("Enable replicated tables")}
```
by default this should be true.

vraghavan78 commented on src/backend/utils/misc/guc_gp.c:
```diff
@@ -2905,6 +2905,17 @@ struct config_bool ConfigureNamesBool_gp[] =
+		 GUC_NO_SHOW_ALL | GUC_NOT_IN_SAMPLE
+		 },
+		 &optimizer_enable_replicated_table,
+		 false,
```
please enable it by true.

chrishajas commented on src/backend/gporca/libgpopt/src/translate/CTranslatorExprToDXL.cpp:
```diff
@@ -6903,6 +6903,18 @@ CTranslatorExprToDXL::GetProperties(const CExpression *pexpr)
 		rows = rows * ulSegments;
 	}
 
+	if (GPOS_FTRACE(EopttraceEnableReplicatedTable) == false &&
```
We might be able to put this logic in `CTranslatorDXLToExpr::Ptabdesc` where we check `if (IMDRelation::EreldistrReplicated == rel_distr_policy)`, which would let us fall back earlier.

dgkimura commented on src/backend/gporca/libgpopt/src/translate/CTranslatorDXLToExpr.cpp:
```diff
@@ -2162,6 +2162,12 @@ CTranslatorDXLToExpr::Ptabdesc(CDXLTableDescr *table_descr)
+		if (GPOS_FTRACE(EopttraceEnableReplicatedTable) == false)
+		{
+			GPOS_RAISE(gpdxl::ExmaMD, gpdxl::ExmiMDObjUnsupported,
+					   GPOS_WSZ_LIT("Replicated tables"));
```
I suggest adding GUC name into error message so it's easy to figure out how to turn the feature on.

dgkimura commented on src/test/regress/sql/rpt.sql:
```diff
@@ -4,6 +4,8 @@
 create schema rpt;
 set search_path to rpt;
 
+set optimizer_enable_replicated_table=true;
```
Why do we need to set it on?  Actually, I suggest adding a quick testcase here that turning GUC off does indeed lead to fallbacks.

thedanhoffman replied:
this line was supposed to be removed. removed it and added the test case you mention at the bottom of the file

dgkimura commented on src/backend/utils/misc/guc_gp.c:
```diff
@@ -2905,6 +2905,17 @@ struct config_bool ConfigureNamesBool_gp[] =
 		NULL, NULL, NULL
 	},
+	{
+		{"optimizer_enable_replicated_table", PGC_USERSET, DEVELOPER_OPTIONS,
```
Actually, I think I take back this comment.  Fallback probably isn't really a query tune change

thedanhoffman commented on src/backend/utils/misc/guc_gp.c:
```diff
@@ -2905,6 +2905,17 @@ struct config_bool ConfigureNamesBool_gp[] =
+		{"optimizer_enable_replicated_table", PGC_USERSET, DEVELOPER_OPTIONS,
+		 gettext_noop("Enable replicated tables"),
+		 NULL,
+		 GUC_NO_SHOW_ALL | GUC_NOT_IN_SAMPLE
```
unsure why I added this. removed it

dgkimura commented on src/test/regress/expected/rpt_optimizer.out:
```diff
@@ -1201,6 +1200,25 @@ explain (costs off) select * from t1_13532 x, t2_13532 y where y.a < random() an
 (8 rows)
 
+-- quick test for optimizer_enable_replicated_table
+explain select * from rep_tab;
+                                  QUERY PLAN                                  
+------------------------------------------------------------------------------
+ Gather Motion 1:1  (slice1; segments: 1)  (cost=0.00..431.00 rows=2 width=4)
+   ->  Seq Scan on rep_tab  (cost=0.00..431.00 rows=6 width=4)
+ Optimizer: Pivotal Optimizer (GPORCA)
+(3 rows)
+
+set optimizer_enable_replicated_table=false;
+explain select * from rep_tab;
+                                           QUERY PLAN                                           
+------------------------------------------------------------------------------------------------
+ Gather Motion 1:1  (slice1; segments: 1)  (cost=10000000001.02..10000000001.02 rows=2 width=4)
+   ->  Seq Scan on rep_tab  (cost=10000000000.00..10000000001.02 rows=2 width=4)
+ Optimizer: Postgres query optimizer
+(3 rows)
```
Because these plan shapes are the same, test may not actually be properly testing fallback.  Suggest setting `optimizer_trace_fallback`.

gpopt reviewed (commented):
Can we check it when translating query to DXL, so that we don't even need a traceflag on Orca side?

chrishajas commented:
> Can we check it when translating query to DXL, so that we don't even need a traceflag on Orca side?

I'm ok with it being here or during translation (we could add it in `CTranslatorRelcacheToDXL::RetrieveRel`. You're right, adding it before we get to Orca means we don't need a traceflag so it's probably a bit better.

thedanhoffman commented:
It doesn't look like any other GUCs are set here and reading from `optimizer_enable_replicated_table` directly always returns the default value. My guess is a new process is spawned and the GUC values aren't transferred. I also removed all Orca traceflag passing with this patch.

dgkimura commented:
On Monday, June 27, 2022 9:50 AM Daniel Hoffman wrote:
> It doesn't look like any other GUCs are set here and reading from optimizer_enable_replicated_table directly always returns the default value. My guess is a new process is spawned and the GUC values aren't transferred.

Can you elaborate a bit more?  GUCs set where and transferred where?  I'm missing the context here.

> I also removed all Orca traceflag passing with this patch.

Did you push these changes?  I don't think the PR reflects this statement yet?

thedanhoffman commented:
@dgkimura Reading the GUC values at the time `CTranslatorRelcacheToDXL` is ran only has the default value for `optimizer_enable_replicated_table`. I should have clarified the latter half: I did both of these changes and did not push either because I couldn't read the GUC value to fall-back to planner.

thedanhoffman commented:
I'm running into an issue that looks like not properly invalidating a cache. The following two queries have different behavior given the current tip of the branch.

This calls `RetrieveRel` once (so the updated value of the GUC is never set, because the cache is not invalidated because of it):
```
drop table if exists foo;
create table foo (a int) distributed replicated;

-- the following calls RetrieveRel once, producing wrong results

set client_min_messages=log;
explain select * from foo;
set optimizer_enable_replicated_table=false;
explain select * from foo;
reset optimizer_enable_replicated_table;
reset client_min_messages;
```

The following calls `RetrieveRel` twice:
```
drop table if exists foo;
create table foo (a int) distributed replicated;

-- the following calls RetrieveRel twice, producing correct results

set client_min_messages=log;
set optimizer_enable_replicated_table=false;
explain select * from foo;
reset optimizer_enable_replicated_table;
explain select * from foo;
reset client_min_messages;
```

I'll look into this further.

thedanhoffman commented:
I'm not pursuing the Relcache fallback method because of the above behavior. I've reverted back to the `CTranslatorDXLToExpr` fallback and added the replicated table GUC as a default in the unit tests

dgkimura commented:
On Tuesday, July 5, 2022 8:51 AM Daniel Hoffman wrote:
> I'm not pursuing the Relcache fallback method because of the above behavior. I've reverted back to the CTranslatorDXLToExpr fallback and added the replicated table GUC as a default in the unit tests

I looked a little closer.  It doesn't work in Relcache because (as I think you mentioned) it caches the metadata object and doesn't retrieve it again during lifetime of the client connection.  But I think @gpopt's suggestion of checking during QueryToDXL can still work if we move check to `CTranslatorUtils::GetTableDescr()` after we retrieved the `IMDRelation` metadata.

thedanhoffman commented:
> I looked a little closer. It doesn't work in Relcache because (as I think you mentioned) it caches the metadata object and doesn't retrieve it again during lifetime of the client connection. But I think @gpopt's suggestion of checking during QueryToDXL can still work if we move check to `CTranslatorUtils::GetTableDescr()` after we retrieved the `IMDRelation` metadata.

Thanks. I'll look into this next

gpopt commented:
Check out the method `NoteDistributionPolicyOpclasses()`.

thedanhoffman commented:
> Check out the method `NoteDistributionPolicyOpclasses(RangeTblEntry)`.

I looked into this but I don't believe it is sufficient. Unless I'm mistaken, the only distribution we have is `intoPolicy` which is the output distribution, so we still operate on the children tables as replicated. The following query should fall back but does not (unless I'm misunderstanding what should be considered for fallback):
```
explain select a.a, b.a from foo a, foo b;

                                    QUERY PLAN
----------------------------------------------------------------------------------
 Gather Motion 1:1  (slice1; segments: 1)  (cost=0.00..1324032.09 rows=1 width=8)
   ->  Nested Loop  (cost=0.00..1324032.09 rows=3 width=8)
         Join Filter: true
         ->  Seq Scan on foo foo_1  (cost=0.00..431.00 rows=3 width=4)
         ->  Seq Scan on foo  (cost=0.00..431.00 rows=3 width=4)
 Optimizer: Pivotal Optimizer (GPORCA)
(6 rows)
```

vraghavan78 commented:
closing this since has be resolved by https://github.com/greenplum-db/gpdb/pull/13763
